flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java

```diff
@@ -74,10 +79,20 @@ private StandaloneJobClusterEntryPoint(
 	}
 
 	@Override
-	protected DispatcherResourceManagerComponentFactory createDispatcherResourceManagerComponentFactory(Configuration configuration) {
+	protected DispatcherResourceManagerComponentFactory createDispatcherResourceManagerComponentFactory(Configuration configuration) throws IOException {
+		final String flinkHomeDir = System.getenv(ENV_FLINK_HOME_DIR);
```

Comment:
1. I have verified it. (Running /test_docker_embedded_job.sh and manually package the per-job docker image and run) . I see this value in `ConfigConstants`. So I thought it was a contract.

1. Since you said this is not a contract, I would like to propose not to rely on `FLINK_HOME`
1.  One possible approach is:
    1. The default behavior is to only look for the UsrLib directory relative to WorkingDir.
    1. If it does not exist, we could try the ENV_FLINK_USR_LIB_DIR directory.

By the way
This requires modifying the working dir in docker-entrypoint.sh to support the default behavior.

Reply:
I think it works in Docker because the Dockerfile defines `FLINK_HOME` as an env variable. The problem is that the other scripts like `flink-daemon.sh` don't do this.

Not sure whether the working directory will work because image people started the cluster from within `FLINK_HOME/bin`.

What do you mean with "We could try the `ENV_FLINK_USR_LIB_DIR` directory first"?

I'm wondering whether introducing a new env variable for every directory is a good practice. Instead one could derive directories like `lib`, `conf`, `plugins`, `usrlib` if `FLINK_HOME` would be defined. So maybe the right approach would be to update the entrypoints to properly set this env variable and derive the other directories relative to this one.

Reply:
>What do you mean with "We could try the ENV_FLINK_USR_LIB_DIR directory first"?

I want to add an env just like `ENV_FLINK_PLUGINS_DIR` and `ENV_FLINK_OPT_DIR`

I think you are right. We might make the view of Flink deployment environment more uniform if we introduce the FLINK_HOME and other directories could derive from it.  So I +1 for your proposal.

I could add a util class such as` DeploymentUtil#getFlinkHome()`to get FLINK_HOME

>The problem is that the other scripts like flink-daemon.sh don't do this.

Yes. Maybe we could export FLINK_HOME env in config.sh to let all the scripts to set the FLINK_HOME env. (I am not pretty sure this because scripts related to pyhton might have a little different. I need to figure out.) 
The other situation such as 'YARN/Kubernate' also should export such env.

Reply:
Yes, if we say that we are using `FLINK_HOME`, then every mode should properly set/export it. One option to enforce this it add a check statement to every Java entrypoint.

Nit: Maybe we could name it `ClusterEntrypointUtils#getFlinkHome` or `ClusterEntrypointUtils#getUserLibDirectory()`.

Reply:
We could add the check state to the  `JobClusterEntrypoint` or  `ClusterEntrypoint`
`checkState(getFlinkHome.isPresent(), "the FLINK_HOME is not set correctly")`

This could avoid adding the check state to every Java entrypoint.

Reply:
Adding the check statement to `ClusterEntrypoint` might break the user who has their own `XXXClusterEntrypoint`.
So I think we should add the check statement to every Java entrypoint.

Reply:
There are also the `MesosTaskExecutorRunner`, `YarnTaskExecutorRunner` adn `TaskManagerRunner`. Ok this looks a bit like a rabbit hole. I would suggest to do the following. Let's derive `usrlib` from `ENV_FLINK_LIB_DIR` in a utility where we can change the implementation easily. Then let's create an issue to pull this thing straight. With this I mean to export `FLINK_HOME`.

Reply:
ok. I could open a Jira to follow this.